vermouth/processors/apply_rubber_band.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2018 University of Groningen
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+import numpy as np
+import networkx as nx
+from .processor import Processor
+from .. import selectors
+
+
+def self_distance_matrix(coordinates):
+    return np.sqrt(((coordinates[:, np.newaxis, :] - coordinates[np.newaxis, :, :]) ** 2).sum(axis=-1))
+
+
+def compute_decay(distance_matrix, shift, rate, power):
+    return np.exp(-rate * ((distance_matrix - shift) **  power))
+
+
+def compute_force_constants(distance_matrix, lower_bound, upper_bound,
+                            decay_factor, decay_power, base_constant,
+                            minimum_force):
+    constants = compute_decay(distance_matrix, lower_bound, decay_factor, decay_power)
+    np.fill_diagonal(constants, 0)
+    constants *= base_constant
+    constants[constants < minimum_force] = 0
+    constants[distance_matrix > upper_bound] = 0
+    return constants
+
+
+def build_connectivity_matrix(graph, separation, selection=None):
+    if separation <= 0:
+        raise ValueError('Separation has to be strictly positive.')
+    if separation == 1:
+        return nx.to_numpy_matrix(graph, nodelist=selection).astype(bool)
+    subgraph = graph.subgraph(selection)
+    connectivity = np.zeros((len(subgraph), len(subgraph)), dtype=bool)
+    for (i, key_i),  (j, key_j) in itertools.combinations(enumerate(subgraph.nodes), 2):
+        shortest_path = len(nx.shortest_path(subgraph, key_i, key_j))
```

Comment:
I would've done it the other way around: make a complete distance matrix, and do numpy magic to find the node_keys between which the distance is what you want. There's not really a convenient method in networkx for this though. Maybe `networkx.algorithms.shortest_paths.dense.floyd_warshall_numpy`, but it would require profiling.